frontend/src/modules/dataset/dataset-browser.tsx

```diff
@@ -0,0 +1,389 @@
+  limitations under the License.
+*/
+import { Box, BreadcrumbGroup, Header, Pagination, SpaceBetween, Table } from '@cloudscape-design/components';
+import React, { useEffect, useMemo } from 'react';
+import { DatasetContext, datasetFromS3Uri } from '../../shared/util/dataset-utils';
+import { getDatasetsList } from '../../entities/dataset/dataset.reducer';
+import { useAppDispatch } from '../../config/store';
+import { useParams } from 'react-router-dom';
+import { modeForDatasetContext as displayModeForDatasetContext, getSelectionType, prefixForPath, resourceForPath, tablePropertiesForDisplayMode } from './dataset-browser.utils';
+import { breadcrumbFromDataset } from './dataset-browser.utils';
+import { DatasetActionType, DatasetBrowserState, DatasetResource, datasetBrowserReducer, getDatasetContents } from './dataset-browser.reducer';
+import Condition from '../condition';
+import { DatasetBrowserDisplayMode, DatasetBrowserManageMode } from './dataset-browser.types';
+import { DatasetType, IDataset } from '../../shared/model';
+import { isFulfilled } from '@reduxjs/toolkit';
+import { MLSTextFilter } from '../textfilter/textfilter';
+import { upperFirst } from 'lodash';
+import { getMatchesCountText } from '../../shared/util/table-utils';
+import { EmptyState } from '../table';
+import { DatasetBrowserProps } from './dataset-browser.types';
+import NotificationService from '../../shared/layout/notification/notification.service';
+import { useUsername } from '../../shared/util/auth-utils';
+
+export function DatasetBrowser (props: DatasetBrowserProps) {
+    const username = useUsername();
+    const dispatch = useAppDispatch();
+    const { projectName } = useParams();
+    const { actions, resource, header, selectableItemsTypes = [], manageMode, onItemsChange } = props;
+    const isSelectingObject = selectableItemsTypes.indexOf('objects') > -1;
+    const isSelectingPrefix = selectableItemsTypes.indexOf('prefixes') > -1;
+    const [state, setState] = React.useReducer(datasetBrowserReducer, {
+        datasetContext: undefined,
+        breadcrumbs: [],
+        items: [],
+        selectedItems: [],
+        manageMode,
+        isLoading: false,
+        nextToken: undefined,
+        username: username!,
+        projectName: projectName!,
+        filter: {
+            filteringText: '',
+            filteringTextDisplay: '',
+            filteredItems: []
+        },
+        pagination: {
+            currentPageIndex: 1,
+            pagesCount: 1,
+            openEnd: false,
+            disabled: false,
+        },
+        refresh: false
+    });
+
+    const displayMode = displayModeForDatasetContext(state.datasetContext);
+    const isCreating = state.manageMode === DatasetBrowserManageMode.Create;
+
+    // memoize notification service so it doesn't cause infinite loops for fetchDatasetResources & fetchDatasets
+    const notificationService = useMemo(() => {
```

Comment:
I'll update the comments but it is because it is used in react hooks. Memoizing it prevents it from being recreated every time which would trigger the other react hooks where it is used.